Port libpas to Windows

iangrunert opened the pull request:
#### f34b2de8939ea29e1f46d834319376219ee22582
<pre>
Port libpas to Windows
<a href="https://bugs.webkit.org/show_bug.cgi?id=283311">https://bugs.webkit.org/show_bug.cgi?id=283311</a>

Reviewed by Yusuke Suzuki.

Introduced pas_thread as a wrapper around pthread, where we
introduce pthread shims for Windows.

Turned on compiler flag mcx16 so it&apos;ll use the cmpxchg16b instruction.
There hasn&apos;t been a x64 CPU built without that feature for 20 years or
so, feels like it&apos;s safe to turn that on.

Currently using fiber-local-storage instead of thread-local-storage.
There&apos;s a performance penalty, but it&apos;s simpler to implement and we can
switch to thread-local-storage in a follow-up PR.

We&apos;re manually committing the memory in StructureAlignedMemoryAllocator.
This is to workaround <a href="https://bugs.webkit.org/show_bug.cgi?id=292771">https://bugs.webkit.org/show_bug.cgi?id=292771</a> -
pas_page_malloc commit isn&apos;t getting called in this codepath. This
shouldn&apos;t use the physical memory until the pages are accessed anyway,
so it should be fine for it to work this way.

Canonical link: <a href="https://commits.webkit.org/295497@main">https://commits.webkit.org/295497@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6910b2d735390a5c04b98a9b6bdd3b3010c8ea17

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104676 "26 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24388 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14809 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109890 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55350 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24779 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32934 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79506 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107682 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19282 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94478 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59813 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19048 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12553 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54725 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/97357 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88745 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12603 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/postmessage-to-client-message-queue.https.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112285 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/103294 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31841 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23432 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88580 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32205 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90707 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88196 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33091 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10861 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27154 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17071 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31766 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37118 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/127573 "Build was cancelled. Recent messages:Printed configuration") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31558 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/127573 "Build was cancelled. Recent messages:Printed configuration") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34899 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33117 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/bmalloc/bmalloc/IsoMallocFallback.cpp:
```diff
@@ -38,6 +38,10 @@ MallocFallbackState mallocFallbackState;
 namespace {
 
+#if BOS(WINDOWS)
+#define strcasecmp _stricmp
```
TIL

Constellation commented on Source/bmalloc/bmalloc/DebugHeap.cpp:
```diff
@@ -122,7 +122,11 @@ void* DebugHeap::malloc(size_t size, FailureAction action)
+    #if !BOS(WINDOWS)
     void* result = ::aligned_alloc(alignment, size);
+    #else
+    void* result = ::_aligned_malloc(alignment, size);
```
Note that it does not work: the memory needs to be deallocated via _aligned_free in Windows and we cannot use free. So it is not possible to implement the same interface with Windows default library. So on Windows, probably we should disable system malloc fallback mechanism

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_commit_mode.h:
```diff
@@ -32,7 +32,7 @@ PAS_BEGIN_EXTERN_C;
 /* pas_commit_mode is a named "is committed" boolean. */
 
-enum pas_commit_mode {
+enum pas_commit_mode : unsigned int {
```
This is C++ one, and libpas is C.

iangrunert replied:
Removed and fixed this problem another way - though this is valid in C23 whenever WebKit upgrades to that.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_log.c:
```diff
@@ -31,7 +31,13 @@
 #include "pas_snprintf.h"
+#if PAS_OS(WINDOWS)
+#include <io.h>
+#define write _write
```
Instead of using macro, we should define a static function and inside it we should switch `write` or `_write`.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_mmap_capability.h:
```diff
@@ -34,7 +34,7 @@ PAS_BEGIN_EXTERN_C;
    on the memory that they ask libpas to manage. This enumeration controls what memory management functions
    libpas can call for some memory that it manages. */
-enum pas_mmap_capability {
+enum pas_mmap_capability : unsigned int {
```
This is C++ thing, not OK in libpas.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_physical_memory_synchronization_style.h:
```diff
@@ -30,7 +30,7 @@
 PAS_BEGIN_EXTERN_C;
 
-enum pas_physical_memory_synchronization_style {
+enum pas_physical_memory_synchronization_style : unsigned int {
```
Ditto.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_utils.h:
```diff
@@ -945,6 +948,9 @@ static inline bool pas_compare_and_swap_pair_weak(void* raw_ptr,
 PAS_IGNORE_WARNINGS_BEGIN("atomic-alignment")
+    #if PAS_CPU(ADDRESS64)
+        PAS_ASSERT((uintptr_t) raw_ptr % 16 == 0); // CMPXCHG16B requires destination be 16-byte aligned
+    #endif
```
This is not correct indentation.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_utils.h:
```diff
@@ -985,6 +991,9 @@ static inline pas_pair pas_compare_and_swap_pair_strong(void* raw_ptr,
 PAS_IGNORE_WARNINGS_BEGIN("atomic-alignment")
+    #if PAS_CPU(ADDRESS64)
+        PAS_ASSERT((uintptr_t) raw_ptr % 16 == 0); // CMPXCHG16B requires destination be 16-byte aligned
+    #endif
```
Ditto.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_thread.c:
```diff
@@ -0,0 +1,193 @@
+//
+// However the fiber-local storage APIs do let you register a per-thread destructor. Even 
+// though we don't care about fibers, the API is strictly better and avoids the magic:
+// https://devblogs.microsoft.com/oldnewthing/20191011-00/?p=102989
```
I think probably blink's way is better. The reason is fiber-local-storage is significantly slower than TLS.

iangrunert replied:
Yeah it does look like in single-threaded FlsGetValue is about 2x slower than TlsGetValue in the simple benchmarks I ran.

I think switching to TLS is fiddly enough that I'd rather land that in a future patch. It looks like libpas with fiber-local-storage is roughly on par with the system allocator for performance in the limited benchmarks I ran.

Constellation commented on Source/bmalloc/bmalloc/DebugHeap.cpp:
```diff
@@ -106,6 +106,55 @@ void DebugHeap::dump()
 
+#elif BOS(WINDOWS)
+
+// DebugHeap unimplemented on Windows
```
Let's mention to the reason why (impedance mismatch between free and aligned free)

Constellation commented on Source/bmalloc/libpas/build_and_test.sh:
```diff
@@ -22,7 +22,7 @@
 # THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-. common.sh
+./common.sh
```
This is a bit different meaning, the previous one is source command

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_committed_pages_vector.c:
```diff
@@ -57,6 +57,9 @@ void pas_committed_pages_vector_construct(pas_committed_pages_vector* vector,
     PAS_SYSCALL(mincore(object, size, (unsigned char*)vector->raw_data));
+#elif PAS_OS(WINDOWS)
+    // TODO mincore equivalent
+    PAS_ASSERT(false);
```
Why is it fine?

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_large_sharing_pool.c:
```diff
@@ -531,11 +531,19 @@ static void splat_live_bytes(pas_large_sharing_node* node,
     bool did_overflow;
     switch (free_mode) {
     case pas_free:
+#if PAS_OS(WINDOWS)
```
Why?

iangrunert replied:
`unsigned long` is 4 bytes on Windows and 8 bytes on Mac / Linux. This mismatch between `size_t` and `unsigned long` causes issues. I've replaced this with a cast from `size_t` to `unsigned long long` which should work on all the 64 bit platforms, but there may be issues if this is used on arm32.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_large_sharing_pool.h:
```diff
@@ -51,9 +51,9 @@ typedef struct pas_physical_memory_transaction pas_physical_memory_transaction;
-    pas_mmap_capability mmap_capability : 1;
+    bool is_committed : 1;
+    bool synchronization_style : 1;
+    bool mmap_capability : 1;
```
This change does not look good. I think you should annotate this is non ms struct in clang-cl

iangrunert replied:
Updated to make this change Windows only with some extra comments. gcc_struct isn't available on clang-cl, unfortunately so I think this is about as good as it can get until C23 unlocks typed enumerations.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_thread.c:
```diff
@@ -0,0 +1,206 @@
+/*
```
Also, in general, please follow to libpas / WebKit coding style. Like, we should not put `*` to the variable side for pointers.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_utils.h:
```diff
@@ -29,7 +29,6 @@
 #include "pas_config.h"
 
 /* These need to be included first. */
-#include <pthread.h>
```
As described "These need to be included first.". This removal is not fine.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_utils.h:
```diff
@@ -985,6 +991,9 @@ static inline pas_pair pas_compare_and_swap_pair_strong(void* raw_ptr,
 #elif PAS_COMPILER(CLANG)
 PAS_IGNORE_WARNINGS_BEGIN("atomic-alignment")
+#if PAS_CPU(ADDRESS64)
+    PAS_ASSERT(!((uintptr_t) raw_ptr % 16)); // CMPXCHG16B requires destination be 16-byte aligned
```
Use PAS_TESTING_ASSERT. This is incredibly hot function.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_utils.h:
```diff
@@ -945,6 +948,9 @@ static inline bool pas_compare_and_swap_pair_weak(void* raw_ptr,
 #elif PAS_COMPILER(CLANG)
 PAS_IGNORE_WARNINGS_BEGIN("atomic-alignment")
+#if PAS_CPU(ADDRESS64)
+    PAS_ASSERT(!((uintptr_t) raw_ptr % 16)); // CMPXCHG16B requires destination be 16-byte aligned
```
Use PAS_TESTING_ASSERT. This is incredibly hot function.

Constellation commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -4000,6 +4000,19 @@ static bool isMJSFile(char *filename)
+    return cached;
+}
+#endif
+
```
This is not necessary. Use `WTF::systemPageSize`.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_thread.c:
```diff
@@ -0,0 +1,206 @@
+    PAS_UNUSED_PARAM(attr);
+    PAS_UNUSED_PARAM(arg);
+
+    // Create thread handle
```
Use C-style comment `/* */`

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_utils.h:
```diff
@@ -946,6 +950,9 @@ static inline bool pas_compare_and_swap_pair_weak(void* raw_ptr,
 #elif PAS_COMPILER(CLANG)
 PAS_IGNORE_WARNINGS_BEGIN("atomic-alignment")
+#if PAS_CPU(ADDRESS64)
+    PAS_TESTING_ASSERT(!((uintptr_t) raw_ptr % 16)); // CMPXCHG16B requires destination be 16-byte aligned
```
Remove space after `(uintptr_t)`

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_utils.h:
```diff
@@ -986,6 +993,9 @@ static inline pas_pair pas_compare_and_swap_pair_strong(void* raw_ptr,
 #elif PAS_COMPILER(CLANG)
 PAS_IGNORE_WARNINGS_BEGIN("atomic-alignment")
+#if PAS_CPU(ADDRESS64)
+    PAS_TESTING_ASSERT(!((uintptr_t) raw_ptr % 16)); // CMPXCHG16B requires destination be 16-byte aligned
```
Ditto.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_status_reporter.c:
```diff
@@ -1144,7 +1155,11 @@ static void* status_reporter_thread_main(void* arg)
+        Sleep(pas_status_reporter_period_in_milliseconds);
+#else
         usleep(pas_status_reporter_period_in_microseconds);
+#endif
```
Can we have `pas_sleep` function abstraction?

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_compact_tagged_ptr.h:
```diff
@@ -47,6 +47,7 @@ PAS_BEGIN_EXTERN_C;
         if (ptr < PAS_INTERNAL_MIN_ALIGN) \
             return ptr; \
         offset = ptr - pas_compact_heap_reservation_base; \
+        PAS_ASSERT(offset < ptr); \
```
I don't think this assert is correct. Please remove this.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_segregated_size_directory.h:
```diff
@@ -54,7 +54,7 @@ PAS_DEFINE_COMPACT_ATOMIC_PTR(pas_segregated_size_directory_data,
 #define PAS_SEGREGATED_SIZE_DIRECTORY_ALIGNMENT_SHIFT_BITS 5
 
-struct pas_segregated_size_directory {
+struct PAS_ALIGNED(sizeof(pas_versioned_field)) pas_segregated_size_directory {
```
Why is this necessary?

iangrunert replied:
https://webkit.slack.com/archives/CU84Q46JZ/p1740702382574929

Without specifying alignment of pas_segregated_size_directory, pas_segregated_size_directory_get_bitfit_size_class returns a non-16-byte-aligned pas_bitfit_size_class pointer.

With the incorrect alignment, it later crashes on movaps and CMPXCHG16B instructions.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_status_reporter.c:
```diff
@@ -61,10 +61,13 @@
+#endif
 
 unsigned pas_status_reporter_enabled = 0;
+unsigned pas_status_reporter_period_in_milliseconds = 10 * 1000;
```
We do not need this.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_status_reporter.c:
```diff
@@ -1135,7 +1138,27 @@ void pas_status_reporter_dump_everything(pas_stream* stream)
     pas_status_reporter_dump_expendable_memories(stream);
 }
 
+void pas_sleep(unsigned microseconds);
```
Remove this.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_status_reporter.c:
```diff
@@ -1135,7 +1138,27 @@ void pas_status_reporter_dump_everything(pas_stream* stream)
+#if PAS_OS(WINDOWS)
+#define R_NULL 0
+
+void pas_sleep(unsigned microseconds)
```
Put `static`.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_status_reporter.c:
```diff
@@ -1135,7 +1138,27 @@ void pas_status_reporter_dump_everything(pas_stream* stream)
+#else
+#define R_NULL NULL
+
+void pas_sleep(unsigned microseconds)
```
Put `static`.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_status_reporter.c:
```diff
@@ -1135,7 +1138,27 @@ void pas_status_reporter_dump_everything(pas_stream* stream)
+    usleep(microseconds);
+}
+
 static void* status_reporter_thread_main(void* arg)
```
This code looks very weird. You should not involve `pas_sleep` definition and this function's signature in one if-def.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_status_reporter.c:
```diff
@@ -1135,7 +1138,27 @@ void pas_status_reporter_dump_everything(pas_stream* stream)
+    Sleep(microseconds);
+}
+
+unsigned status_reporter_thread_main(void* arg)
```
Need `static`.

Constellation commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -4000,6 +4000,13 @@ static bool isMJSFile(char *filename)
+    return WTF::pageSize();
+}
+#endif
+
```
Instead of defining `getpagesize`, we should replace `getpagesize` use with `WTF::pageSize`.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_large_sharing_pool.c:
```diff
@@ -531,11 +531,11 @@ static void splat_live_bytes(pas_large_sharing_node* node,
     switch (free_mode) {
     case pas_free:
-        did_overflow = __builtin_usubl_overflow(node->num_live_bytes, delta, &node->num_live_bytes);
+        did_overflow = __builtin_usubll_overflow((unsigned long long) node->num_live_bytes, delta, (unsigned long long *) &node->num_live_bytes);
```
Let's have `PAS_TESTING_ASSERT(sizeof(size_t) == sizeof(unsigned long long))`. Otherwise, this is wrong.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_large_sharing_pool.c:
```diff
@@ -531,11 +531,11 @@ static void splat_live_bytes(pas_large_sharing_node* node,
         break;
     case pas_allocated:
-        did_overflow = __builtin_uaddl_overflow(node->num_live_bytes, delta, &node->num_live_bytes);
+        did_overflow = __builtin_uaddll_overflow((unsigned long long) node->num_live_bytes, delta, (unsigned long long *) &node->num_live_bytes);
```
Ditto.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_utils.h:
```diff
@@ -1112,13 +1122,13 @@ static inline unsigned pas_hash_ptr(const void* ptr)
 /* Undefined for value == 0. */
 static inline unsigned pas_log2(uintptr_t value)
 {
-    return (sizeof(uintptr_t) * 8 - 1) - (unsigned)__builtin_clzl(value);
```
We should have `PAS_TESTING_ASSERT(sizeof(uintptr_t) == sizeof(unsigned long long))`. Otherwise, this is wrong.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_utils.h:
```diff
@@ -1112,13 +1122,13 @@ static inline unsigned pas_hash_ptr(const void* ptr)
 /* Undefined for value <= 1. */
 static inline unsigned pas_log2_rounded_up(uintptr_t value)
 {
-    return (sizeof(uintptr_t) * 8 - 1) - ((unsigned)__builtin_clzl(value - 1) - 1);
```
We should have `PAS_TESTING_ASSERT(sizeof(uintptr_t) == sizeof(unsigned long long))`. Otherwise, this is wrong.

iangrunert committed:
Port libpas to Windows
https://bugs.webkit.org/show_bug.cgi?id=283311

Reviewed by Yusuke Suzuki.

Introduced pas_thread as a wrapper around pthread, where we
introduce pthread shims for Windows.

Turned on compiler flag mcx16 so it'll use the cmpxchg16b instruction.
There hasn't been a x64 CPU built without that feature for 20 years or
so, feels like it's safe to turn that on.

Currently using fiber-local-storage instead of thread-local-storage.
There's a performance penalty, but it's simpler to implement and we can
switch to thread-local-storage in a follow-up PR.

We're manually committing the memory in StructureAlignedMemoryAllocator.
This is to workaround https://bugs.webkit.org/show_bug.cgi?id=292771 -
pas_page_malloc commit isn't getting called in this codepath. This
shouldn't use the physical memory until the pages are accessed anyway,
so it should be fine for it to work this way.

Canonical link: https://commits.webkit.org/295497@main
